File: packages/go/rest.go
package erpc

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"time"
)

type restTransport struct {
	token    string
	endpoint *url.URL
	headers  http.Header
	timeout  time.Duration
	client   *http.Client
}

func newRESTTransport(token string, endpoint *url.URL, config resolvedConfig) *restTransport {
	return &restTransport{token: token, endpoint: endpoint, headers: config.headers.Clone(), timeout: config.timeout, client: config.httpClient}
}

func (t *restTransport) getJSON(ctx context.Context, path string, query url.Values, result any) error {
	response, err := t.get(ctx, path, query, "application/json")
	if err != nil {
		return err
	}
	defer response.Body.Close()
	if err := json.NewDecoder(io.LimitReader(response.Body, 16<<20)).Decode(result); err != nil {
		return sdkError(ErrorInvalidResponse, "ERPC returned malformed JSON")
	}
	return nil
}

func (t *restTransport) get(ctx context.Context, path string, query url.Values, accept string) (*http.Response, error) {
	u := endpointPath(t.endpoint, path)
	u.RawQuery = query.Encode()
	requestContext, cancel := context.WithTimeout(ctx, t.timeout)
	request, err := http.NewRequestWithContext(requestContext, http.MethodGet, u.String(), nil)
	if err != nil {
		cancel()
		return nil, sdkError(ErrorConfig, "unable to create REST request")
	}
	request.Header = t.headers.Clone()
	request.Header.Set("Authorization", "Bearer "+t.token)
	request.Header.Set("Accept", accept)
	response, err := t.client.Do(request)
	if err != nil {
		cancel()
		if requestContext.Err() != nil {
			return nil, requestContext.Err()
		}
		return nil, sdkError(ErrorTransport, "unable to reach ERPC")
	}
	response.Body = &cancelReadCloser{ReadCloser: response.Body, cancel: cancel}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		response.Body.Close()
		return nil, &Error{Kind: ErrorHTTP, Message: "request failed", Status: response.StatusCode}
	}
	return response, nil
}

type cancelReadCloser struct {
	io.ReadCloser
	cancel context.CancelFunc
}

func (c *cancelReadCloser) Close() error {
	c.cancel()
	return c.ReadCloser.Close()
}

File: packages/rust/src/rest.rs
use std::{future::Future, time::Duration};

use reqwest::{Client, Response, header::HeaderMap};
use serde::de::DeserializeOwned;
use tokio_util::sync::CancellationToken;
use url::Url;

use crate::{ErpcError, Result, error::timeout};

pub(crate) type Query = Vec<(String, String)>;

#[derive(Clone)]
pub(crate) struct RestTransport {
    api_key: String,
    endpoint: Url,
    headers: HeaderMap,
    timeout: Duration,
    client: Client,
    configured: bool,
    namespace: String,
}

impl RestTransport {
    pub fn new(
        api_key: String,
        endpoint: Url,
        headers: HeaderMap,
        timeout: Duration,
        client: Client,
    ) -> Self {
        Self {
            api_key,
            endpoint,
            headers,
            timeout,
            client,
            configured: true,
            namespace: "REST".to_owned(),
        }
    }

    pub fn unavailable(namespace: impl Into<String>, timeout: Duration, client: Client) -> Self {
        Self {
            api_key: String::new(),
            endpoint: Url::parse("http://127.0.0.1/").expect("static unavailable endpoint"),
            headers: HeaderMap::new(),
            timeout,
            client,
            configured: false,
            namespace: namespace.into(),
        }
    }

    pub fn url(&self, path: &str, query: &Query) -> Url {
        let mut url = self.endpoint.clone();
        let base = url.path().trim_end_matches('/');
        url.set_path(&format!("{base}/{}", path.trim_start_matches('/')));
        if !query.is_empty() {
            url.query_pairs_mut().extend_pairs(query.iter());
        }
        url
    }

    pub async fn get<T: DeserializeOwned>(
        &self,
        path: &str,
        query: Query,
        cancellation: Option<&CancellationToken>,
    ) -> Result<T> {
        let operation = async {
            let response = self.send(path, query).await?;
            let body = response
                .text()
                .await
                .map_err(|_| ErpcError::Transport("Unable to read ERPC response".to_owned()))?;
            serde_json::from_str(&body)
                .map_err(|_| ErpcError::InvalidResponse("ERPC returned malformed JSON".to_owned()))
        };
        self.run(operation, cancellation).await
    }

    pub async fn get_response(
        &self,
        path: &str,
        query: Query,
        cancellation: Option<&CancellationToken>,
    ) -> Result<Response> {
        self.run(self.send(path, query), cancellation).await
    }

    async fn send(&self, path: &str, query: Query) -> Result<Response> {
        if !self.configured {
            return Err(ErpcError::NotConfigured(self.namespace.clone()));
        }
        let url = self.url(path, &query);
        let response = self
            .client
            .get(url)
            .headers(self.headers.clone())
            .bearer_auth(&self.api_key)
            .header(reqwest::header::ACCEPT, "application/json")
            .send()
            .await
            .map_err(|_| ErpcError::Transport("Unable to reach ERPC".to_owned()))?;
        if !response.status().is_success() {
            return Err(ErpcError::Http {
                status: response.status().as_u16(),
            });
        }
        Ok(response)
    }

    async fn run<T>(
        &self,
        operation: impl Future<Output = Result<T>>,
        cancellation: Option<&CancellationToken>,
    ) -> Result<T> {
        let timed = tokio::time::timeout(self.timeout, operation);
        let result = if let Some(cancellation) = cancellation {
            tokio::select! {
                biased;
                () = cancellation.cancelled() => return Err(ErpcError::Aborted),
                result = timed => result,
            }
        } else {
            timed.await
        };
        result.map_err(|_| timeout(self.timeout))?
    }
}
